### ISSUE TITLE
(Linux) Enabling UFW causes an IPv6 leak

### ISSUE THREAD

**Issue by jordan-ivpn** [ivpn/desktop-app#93]
# Bug report

## Describe your environment

* Device: _____  
* OS name and version: Manjaro updated to 2021-11-01  
* IVPN app version: 3.3.30 installed from the AUR

## Describe the problem

With the IVPN App connected (any protocol) and the app's firewall enabled, if the Uncomplicated Firewall (UFW) firewall manager is enabled with no added rules, IPv6 leaks from the system.

### Steps to reproduce:

1. Connect the IVPN App, firewall enabled, no IPv6 for WireGuard, check for IPv6 leaks: `curl -6 ifconfig.co`
2. Install the `ufw` package: `yay -S ufw`
3. Enable UFW and check status:`sudo ufw enable && sudo ufw status verbose`
4. Check for leaks: `curl ifconfig.co && curl -6 ifconfig.co`

### Observed Results:

The `curl` lookup in step 1 above fails because the IVPN App's firewall blocks IPv6 traffic.

UFW adds ~135 extra lines to ip6tables (`sudo ip6tables -S`) by default when it is enabled.  This does not include any custom rules that a user adds.

The `curl` lookup in step 4 displays an IPv6 address.

**Note:** There are no leaks with WireGuard when IPv6 is enabled for the VPN connection in the app's Settings > Connection area.

The leak persists after disabling UFW (`sudo ufw disable`), which is likely caused by the ~35 rules it leaves behind.  A system restart offers a heavy-handed way to clear these leftover rules and stop the leak.

The leak is also present using the steps above in this order: 2, 3, 1, 4.

### Expected Results:

There are no IPv6 leaks.

**Comment by sata** [comment]
I discovered something similarly. I observed traffic passing via the wrong chain when using UFW. 

```
Chain INPUT (policy DROP 6405 packets, 1931K bytes)
 pkts bytes target     prot opt in     out     source               destination         
  1M   1G ufw-before-logging-input  all  --  any    any     anywhere             anywhere            
  1M   1G ufw-before-input  all  --  any    any     anywhere             anywhere            
13720 3118K ufw-after-input  all  --  any    any     anywhere             anywhere            
10499 2904K ufw-after-logging-input  all  --  any    any     anywhere             anywhere            
10499 2904K ufw-reject-input  all  --  any    any     anywhere             anywhere            
10499 2904K ufw-track-input  all  --  any    any     anywhere             anywhere            
 2979  913K IVPN-IN    all  --  any    any     anywhere             anywhere            


...
ufw-before-input
1M   1G ACCEPT     all  --  any    any     anywhere             anywhere             ctstate RELATED,ESTABLISHED
```

I believe it's due to the fact that IVPN firewall rules are [appended to the end rather than inserted at the head](https://github.com/ivpn/desktop-app/blob/5be864b428ccfdf5af52ddfdb96c7139ed2578d4/daemon/References/Linux/etc/firewall.sh#L140-L147).

So obviously in an established connection, the chain `ufw-before-input` would match before even reaching `IVPN-IN`. 

However, if ivpn firewall rules were to be inserted at the head, it would also imply IVPN silently ignoring the iptables chains that have been setup by the user which wouldn't be what many expect or wish.

Would be nice to be able to choose what IVPN does, I would prefer insert over append in this case. I'm patching it for myself but would like to find a more general approach.

**Comment by stenya** [comment]
@gorkapernas  v3.8.10 is ready for testing

**Comment by gorkapernas** [comment]
Verified fixed on v3.8.10, Ubuntu 20.04 and Fedora 35, tested with `ufw` as described in the ticket and with native IPV6 traffic. There are no more IPV6 leaks when connected to IVPN and `ufw` is enabled.

**Comment by stenya** [comment]
v3.8.20 released
